Add management command to update/sync usernames

azimpradhan opened the pull request:
Currently, the LMS and Forums each store their own user details. This
caused a problem as the separate databases could fall out of sync; a
user changing their username via the LMS would not see this change
reflected in their existing posts.

This commit creates a new management command rename_user which can be
used to rename a user in the LMS and update any previous forum posts
to reference their new username.

2 full test suites are included, and can be run by:
python ./manage.py lms test --verbosity=1 lms/djangoapps/django_comment_client/tests/test_management_commands_rename_user.py --traceback --settings=test
python ./manage.py lms test --verbosity=1 lms/djangoapps/django_comment_client/tests/test_management_utils.py --traceback --settings=test

azimpradhan commented:
@stvstnfrd 
Here is the rebased version of the commit. This is the original username rename tool with the added configuration settings for MongoClient. In order for this to work on production you will need to create a FORUM_MONGO_PARAMS hash that uses data from the secrets repo.

Also, can can you please look at this PR, which adds the rename tool to the UI. I modeled it based on other tools in the sysadmin dashboard:

stvstnfrd commented:
Thanks for resubmitting this @azimpradhan . In the future, if you're resubmitting a PR or adding code changes following a review, could you hold off on squashing until the end?

When things are split into separate commits, it's easy for a reviewer to see that there is only a small delta since they last checked the code and they can focus only on the new bits. When the code is squashed, it either necessitates a complete re-review or it runs the risk of missing mistakes.

In this case, I'm noticing at least one minor difference between this version and the last (that doesn't look like an intentional design change), which I'll point out in-line.

azimpradhan commented:
Yes, makes sense. I will hold of on squashing until the end next time.

stvstnfrd commented:
Actually, I guess I was wrong on what I thought had changed, though that underscores the importance of separating the diffs :P

stvstnfrd commented on lms/djangoapps/django_comment_client/management/commands/rename_user.py:
```diff
@@ -0,0 +1,52 @@
+                old_username=args[0]
+            ))
+        else:
+            print "Changed username of user: {old_username} to {new_username}".format(
```
I could have _sworn_ this was using the print _function_ instead of the print _statement_ at one point...

But apparently, I'm wrong [1].

[1] https://github.com/Stanford-Online/edx-platform/pull/187/files#diff-61d027f9610dfde92401417cc6074c38R49

stvstnfrd commented on lms/djangoapps/django_comment_client/management_utils.py:
```diff
@@ -0,0 +1,50 @@
+
+MONGO_PARAMS = settings.FORUM_MONGO_PARAMS
+
+CREDENTIALS = ''
```
This block here looks to be repeated below in our tests. Can we refactor it into a common helper method? Maybe:

```python
def get_mongo_connection_string():
    # ...
```

stvstnfrd commented on lms/djangoapps/django_comment_client/tests/test_management_commands_rename_user.py:
```diff
@@ -0,0 +1,148 @@
+
+MONGO_PARAMS = settings.FORUM_MONGO_PARAMS
+
+CREDENTIALS = ''
```
See note above about creating a helper method.

```python
from django_comment_client.management_utils import get_mongo_connection_string
```

stvstnfrd commented on lms/djangoapps/django_comment_client/tests/test_management_utils.py:
```diff
@@ -0,0 +1,260 @@
+
+MONGO_PARAMS = settings.FORUM_MONGO_PARAMS
+
+CREDENTIALS = ''
```
See note above about helper method.

stvstnfrd commented on lms/envs/aws.py:
```diff
@@ -490,3 +490,6 @@
 ##### METRICS DATA SOURCE #####
 MAX_ENROLLEES_FOR_METRICS_USING_DB = ENV_TOKENS.get('MAX_ENROLLEES_FOR_METRICS_USING_DB', MAX_ENROLLEES_FOR_METRICS_USING_DB)
+
+####################### Forum mongo credentials #######################
```
Heh, I know you're just following along with the existing convention here with the big bunch of leading `#`'s, but it actually turns out to be a Pylint violation. Naturally, we shouldn't mess with the existing lines, but how do you feel about simplifying this to:

```python
# Forum mongo credentials
```

stvstnfrd commented:
To make sure I'm looking at this correctly, I think the only bits that changed are:
- the creation of the `CONNECTION_STRING`
- the calls to `MongoClient` now pass in the `CONNECTION_STRING`

Was there anything else?

azimpradhan commented:
No, it was just creating CONNECTION_STRING, calling MongoClient with it, and adding the appropriate hashes to aws and to common.py
By the way, @stvstnfrd I am planning to add 
def get_mongo_connection_string():
to helpers.py under django_comment_client
does that sound good?

stvstnfrd commented:
Hrmm, I think my preference would be to stick it in `management_utils.py`. My line of thinking there is everything else in this PR is added in a new file, instead of modifying existing files (except the settings files, where are hands are tied). By keeping everything in new files (files which don't exist upstream), we make it much less likely that we'll encounter conflicts when we merge from upstream.

Let me know if you have strong feelings for putting it in `helpers.py`

azimpradhan commented:
Okay @stvstnfrd I stuck the helper in management_utils.py as you suggested and updated the PR. Tests passed. My changes in a separate commit (so that it is easier to see exactly what I changed). Please review and let me know when I should squash. Thanks!

stvstnfrd commented:
@azimpradhan Ship it!

I have made the corresponding changes for the authentication; I'm going to PR those after you've squashed and merged this, so that I can reference the commit hash in my PR.

Great work!

stvstnfrd commented:
Could you take a look at the Mongo-related test failure before merging?

azimpradhan commented:
retest this please
